### PR TITLE
WIP [multicast_dns]migrate to null-safety

### PR DESCRIPTION
**Pull request by Sameerkash**
Migrate `multicast_dns` to null-sfety

This is related to flutter/flutter#75567

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated `CHANGELOG.md` to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [ ] All existing and new tests are passing.

### CODE REVIEW

File: packages/multicast_dns/test/lookup_resolver_test.dart [hunk 1]
```diff
@@ -59,8 +59,8 @@ void testResult2() {
     final ResourceRecord response2 =
         ip4Result('yyy.local', InternetAddress('2.3.4.5'));
     resolver.handleResponse(<ResourceRecord>[response2, response1]);
-    final IPAddressResourceRecord result1 = await futureResult1.first;
-    final IPAddressResourceRecord result2 = await futureResult2.first;
+    final IPAddressResourceRecord result1 = await (futureResult1.first as Future<IPAddressResourceRecord>);
```
Comment (Sameerkash) [inline packages/multicast_dns/test/lookup_resolver_test.dart:62]: Confusing typecast, the test fails with 

`IPAddressResourceRecord extends ResourceRecord.`

`type 'Future<ResourceRecord>' is not a subtype of type 'Future<IPAddressResourceRecord>' in type cast`

How can this be resolved?

File: packages/multicast_dns/test/client_test.dart [hunk 2]
```diff
@@ -10,11 +10,11 @@ import 'package:test/test.dart';
 
 void main() {
   test('Can inject datagram socket factory and configure mdns port', () async {
-    int lastPort;
+    int? lastPort;
     final MockRawDatagramSocket mockRawDatagramSocket = MockRawDatagramSocket();
```
Comment (Sameerkash) [inline packages/multicast_dns/test/client_test.dart:14]: `mockRawDatagramSocket.address` is Null somehow, not sure why. 

test fails with `Null is not a subtype of type  InternetAddress`

**Review by Sameerkash — COMMENTED** [review]
@stuartmorgan  Could you please take a look at this?

File: packages/multicast_dns/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,6 @@
+## 1.0.0-nullsafety
+* Migrate to null safety
```
Comment (stuartmorgan) [inline packages/multicast_dns/CHANGELOG.md:2]: 0.3.0, rather than 1.0.0

File: packages/multicast_dns/test/client_test.dart (hunk 2 above)
Comment (stuartmorgan) [reply packages/multicast_dns/test/client_test.dart:14]: Mockito doesn't mix very well with nullability; see https://github.com/dart-lang/mockito/blob/master/NULL_SAFETY_README.md

How hard would this class be to mock manually?

**Review by stuartmorgan — COMMENTED** [review]
@blasten It looks like you've done some maintenance work on this package; do you feel familiar enough with it to do the null-safety migration review?

File: packages/multicast_dns/test/client_test.dart (hunk 2 above)
Comment (Sameerkash) [reply packages/multicast_dns/test/client_test.dart:14]: Let me try

File: packages/multicast_dns/lib/src/lookup_resolver.dart [hunk 4]
```diff
@@ -25,7 +25,7 @@ class PendingRequest extends LinkedListEntry<PendingRequest> {
   final StreamController<ResourceRecord> controller;
 
   /// The timer for the request.
-  Timer timer;
+  late Timer timer;
```
Comment (blasten) [inline packages/multicast_dns/lib/src/lookup_resolver.dart:28]: maybe make `_timer` `nullable` instead.  This way, a consumer of this class can check if the value is already set, etc..

File: packages/multicast_dns/test/lookup_resolver_test.dart (hunk 1 above)
Comment (blasten) [reply packages/multicast_dns/test/lookup_resolver_test.dart:62]: what about `(await futureResult1.first) as IPAddressResourceRecord`?

File: packages/multicast_dns/test/client_test.dart (hunk 2 above)
Comment (Sameerkash) [reply packages/multicast_dns/test/client_test.dart:14]: Trying to wrap my head around this.

Comment (Sameerkash) [reply packages/multicast_dns/test/client_test.dart:14]: will try to implement the manual way.

File: packages/multicast_dns/test/lookup_resolver_test.dart (hunk 1 above)
Comment (Sameerkash) [reply packages/multicast_dns/test/lookup_resolver_test.dart:62]: Nope this isn't working.

**Review by Sameerkash — COMMENTED** [review]
cc @blasten

**Comment by stuartmorgan** [comment]
Closing in favor of https://github.com/flutter/packages/pull/297 since it's a complete migration, and this is still WIP. (Unfortunately it looks like Jonah didn't see that this was in-progress)

**Pull request closed without merging**